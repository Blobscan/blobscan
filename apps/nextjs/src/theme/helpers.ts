import type { SystemProps } from "@chakra-ui/styled-system";

/**
 * Styles to disable the native focus ring on chakra components.
 * See https://github.com/chakra-ui/chakra-ui/issues/708
 */
export function chakraFocusRingDisabledStyles(): SystemProps {
  return {
    _focusVisible: { boxShadow: "none", outline: "none" },
  };
}
