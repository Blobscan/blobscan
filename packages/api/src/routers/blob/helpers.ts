import type {
  BlobscanPrismaClient,
  EthUsdPrice,
  ExtendedBlobDataStorageReferenceSelect,
  Prisma,
} from "@blobscan/db";
import { EthUsdPriceModel } from "@blobscan/db/prisma/zod";
import { z } from "@blobscan/zod";

import type {
  ExpandedBlob,
  ExpandedBlock,
  ExpandedTransaction,
  Expands,
} from "../../middlewares/withExpands";
import type { Prettify } from "../../types";
import { isFullyDefined } from "../../utils";
import {
  normalizePrismaBlobDataStorageReferencesFields,
  normalizePrismaTransactionFields,
} from "../../utils";
import {
  baseBlobSchema,
  baseBlockSchema,
  baseTransactionSchema,
  prismaBlobOnTransactionSchema,
} from "../../zod-schemas";

const blobDataStorageReferenceSelect = {
  blobStorage: true,
  url: true,
} satisfies ExtendedBlobDataStorageReferenceSelect;

export const extendedBlobDataStorageReferenceArgs: Prisma.Args<
  BlobscanPrismaClient["blobDataStorageReference"],
  "findFirstOrThrow"
> = {
  select: blobDataStorageReferenceSelect,
  orderBy: {
    blobStorage: "asc",
  },
};

export const baseBlobSelect = {
  commitment: true,
  proof: true,
  size: true,
  usageSize: true,
  versionedHash: true,
} satisfies Prisma.BlobSelect;

export const baseBlobOnTransactionSelect = {
  blobHash: true,
  blockHash: true,
  blockNumber: true,
  blockTimestamp: true,
  index: true,
  txHash: true,
  txIndex: true,
} satisfies Prisma.BlobsOnTransactionsSelect;

export type ExtendedBlobDataStorageReference = NonNullable<
  Prisma.Result<
    BlobscanPrismaClient["blobDataStorageReference"],
    { select: typeof blobDataStorageReferenceSelect },
    "findFirst"
  >
>;
type PrismaBlob = Prisma.BlobGetPayload<{
  select: typeof baseBlobSelect;
}> & {
  dataStorageReferences: ExtendedBlobDataStorageReference[];
};

type PrismaBlobOnTransaction = Prisma.BlobsOnTransactionsGetPayload<{
  select: typeof baseBlobOnTransactionSelect;
}>;

export type CompletedPrismaBlob = Prettify<
  PrismaBlob & {
    transactions: Prettify<
      PrismaBlobOnTransaction & {
        block?: ExpandedBlock;
        transaction?: ExpandedTransaction & {
          block: { blobGasPrice: ExpandedBlock["blobGasPrice"] };
        };
      }
    >[];
  }
>;

export type CompletedPrismaBlobOnTransaction = Prettify<
  PrismaBlobOnTransaction & {
    blob: {
      dataStorageReferences: ExtendedBlobDataStorageReference[];
    } & ExpandedBlob;
    block?: ExpandedBlock;
    transaction?: ExpandedTransaction & {
      block: { blobGasPrice: ExpandedBlock["blobGasPrice"] };
    };
  }
>;
export function createBlobSelect(expands: Expands) {
  const blockExpand = expands.block ? { block: expands.block } : {};
  const txExpand = expands.transaction
    ? {
        transaction: {
          select: {
            ...expands.transaction.select,
            block: {
              select: {
                blobGasPrice: true,
              },
            },
          },
        },
      }
    : {};

  return {
    ...baseBlobSelect,
    dataStorageReferences: extendedBlobDataStorageReferenceArgs,
    transactions: {
      select: {
        ...baseBlobOnTransactionSelect,
        ...blockExpand,
        ...txExpand,
      },
      orderBy: {
        blockNumber: "desc",
      },
    },
  } satisfies Prisma.BlobSelect;
}

export function createBlobsOnTransactionsSelect(expands: Expands) {
  const blockExpand = expands.block ? { block: expands.block } : {};
  const txExpand = expands.transaction
    ? {
        transaction: {
          select: {
            ...expands.transaction.select,
            block: {
              select: {
                blobGasPrice: true,
              },
            },
          },
        },
      }
    : {};

  return {
    ...baseBlobOnTransactionSelect,
    blob: {
      select: {
        ...baseBlobSelect,
        dataStorageReferences: extendedBlobDataStorageReferenceArgs,
      },
    },
    ...txExpand,
    ...blockExpand,
  } satisfies Prisma.BlobsOnTransactionsSelect;
}

const responseBlobOnTxSchema = prismaBlobOnTransactionSchema
  .omit({ blobHash: true })
  .required({
    blockHash: true,
    blockNumber: true,
    blockTimestamp: true,
    index: true,
    txHash: true,
    txIndex: true,
  });

export const responseBlobSchema = baseBlobSchema.extend({
  transactions: z.array(
    responseBlobOnTxSchema
      .merge(
        baseTransactionSchema
          .omit({
            hash: true,
            blockNumber: true,
            blockTimestamp: true,
            blockHash: true,
          })
          .partial()
      )
      .extend({
        block: baseBlockSchema
          .omit({
            hash: true,
            number: true,
            timestamp: true,
          })
          .partial({
            blobGasBaseUsdFee: true,
            blobGasUsdPrice: true,
          })
          .optional(),
        ethUsdPrice: EthUsdPriceModel.shape.price.optional(),
      })
  ),
});

export const responseBlobOnTransactionSchema = responseBlobOnTxSchema
  .merge(baseBlobSchema)
  .extend({
    block: baseBlockSchema
      .omit({
        hash: true,
        number: true,
        timestamp: true,
      })
      .optional(),
    transaction: baseTransactionSchema
      .omit({
        hash: true,
        blockNumber: true,
        blockTimestamp: true,
        blockHash: true,
      })
      .optional(),
  });

export type ResponseBlob = z.input<typeof responseBlobSchema>;
export type ResponseBlobOnTransaction = z.input<
  typeof responseBlobOnTransactionSchema
>;

export function toResponseBlob(
  {
    dataStorageReferences,
    transactions: blobOnTxs,
    ...prismaBlob
  }: CompletedPrismaBlob,
  ethUsdPrices: EthUsdPrice["price"][]
): ResponseBlob {
  const responseBlob: ResponseBlob = {
    ...prismaBlob,
    dataStorageReferences: normalizePrismaBlobDataStorageReferencesFields(
      dataStorageReferences
    ),
    transactions: blobOnTxs.map(
      (
        {
          block: { computeUsdFields, ...restBlock } = {},
          transaction: prismaTx = {},
          ...restBlobOnTx
        },
        i
      ) => {
        const txEthUsdPrice = ethUsdPrices[i];

        const blockUsdFields =
          txEthUsdPrice && computeUsdFields
            ? computeUsdFields(txEthUsdPrice)
            : undefined;
        const txBlock = {
          ...restBlock,
          ...(blockUsdFields ?? {}),
        };

        let responseBlobOnTx: ResponseBlob["transactions"][number] = {
          ...restBlobOnTx,
          ...txBlock,
          ethUsdPrice: txEthUsdPrice?.toNumber(),
        };

        if (isFullyDefined(prismaTx)) {
          const {
            block,
            decodedFields,
            from,
            fromId,
            toId,
            computeBlobGasBaseFee,
            computeUsdFields,
            ...restPrismaTx
          } = prismaTx;

          responseBlobOnTx = {
            ...responseBlobOnTx,
            ...restPrismaTx,
            ...normalizePrismaTransactionFields({
              decodedFields,
              from,
              fromId,
              toId,
            }),
            blobGasBaseFee: computeBlobGasBaseFee(block.blobGasPrice),
            ...(txEthUsdPrice
              ? computeUsdFields({
                  blobGasPrice: block.blobGasPrice,
                  ethUsdPrice: txEthUsdPrice,
                })
              : {}),
          };
        }

        return responseBlobOnTx;
      }
    ),
  };

  return responseBlob;
}

export function toResponseBlobOnTransaction(
  prismaBlobOnTransaction: CompletedPrismaBlobOnTransaction,
  ethUsdPrice?: EthUsdPrice["price"]
): ResponseBlobOnTransaction {
  const {
    blobHash,
    blob: { dataStorageReferences, ...blob },
    transaction,
    block,
    ...restBlobOnTransaction
  } = prismaBlobOnTransaction;

  const responseBlobOnTx: ResponseBlobOnTransaction = {
    ...blob,
    versionedHash: blobHash,
    dataStorageReferences: normalizePrismaBlobDataStorageReferencesFields(
      dataStorageReferences
    ),
    ...restBlobOnTransaction,
  };

  if (transaction) {
    const {
      block: txBlock,
      decodedFields,
      from,
      fromId,
      toId,
      computeBlobGasBaseFee,
      computeUsdFields,
      ...restTx
    } = transaction;

    responseBlobOnTx.transaction = {
      ...restTx,
      blobGasPrice: txBlock.blobGasPrice,
      ...normalizePrismaTransactionFields({
        decodedFields,
        from,
        fromId,
        toId,
      }),
      blobGasBaseFee: computeBlobGasBaseFee(txBlock.blobGasPrice),
      ...(ethUsdPrice
        ? computeUsdFields({
            blobGasPrice: txBlock.blobGasPrice,
            ethUsdPrice,
          })
        : {}),
    };
  }

  if (block) {
    const { computeUsdFields, ...restBlock } = block;
    responseBlobOnTx.block = {
      ...restBlock,
      ...(ethUsdPrice ? computeUsdFields(ethUsdPrice) : {}),
    };
  }

  return responseBlobOnTx;
}
