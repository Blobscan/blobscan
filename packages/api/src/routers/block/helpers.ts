import type { BlobStorageManager } from "@blobscan/blob-storage-manager";
import type { BlobscanPrismaClient, Prisma } from "@blobscan/db";
import { z } from "@blobscan/zod";

import type {
  ExpandedBlob,
  ExpandedTransaction,
  Expands,
} from "../../middlewares/withExpands";
import type { Filters } from "../../middlewares/withFilters";
import type { Prettify } from "../../types";
import { isFullyDefined } from "../../utils";
import { retrieveBlobData } from "../../utils";
import {
  deriveTransactionFields,
  normalizePrismaBlobFields,
  normalizePrismaTransactionFields,
} from "../../utils/transformers";
import {
  baseBlobSchema,
  baseBlockSchema,
  baseTransactionSchema,
} from "../../zod-schemas";

const prismaBlockSelect = {
  hash: true,
  number: true,
  timestamp: true,
  slot: true,
  blobGasUsed: true,
  blobAsCalldataGasUsed: true,
  blobGasPrice: true,
  excessBlobGas: true,
} satisfies Prisma.BlockSelect;

export type PrismaBlock = Prisma.BlockGetPayload<{
  select: typeof prismaBlockSelect;
}>;

export type CompletePrismaBlock = Prettify<
  PrismaBlock & {
    transactions: Prettify<
      {
        hash: string;
      } & Partial<ExpandedTransaction> & {
          blobs: Prettify<
            { blobHash: string } & {
              blob?: ExpandedBlob;
            }
          >[];
        }
    >[];
  }
>;

export function createBlockSelect(expands: Expands, filters?: Filters) {
  const blobExpand = expands.blob ? { blob: expands.blob } : {};
  const transactionExpand = expands.transaction?.select ?? {};

  return {
    ...prismaBlockSelect,
    transactions: {
      select: {
        hash: true,
        ...transactionExpand,
        blobs: {
          select: {
            blobHash: true,
            ...blobExpand,
          },
          orderBy: {
            index: "asc",
          },
        },
      },
      orderBy: {
        index: "asc",
      },
      where: filters?.transactionFilters,
    },
  } satisfies Prisma.BlockSelect;
}

export const responseBlockSchema = baseBlockSchema.extend({
  transactions: z.array(
    baseTransactionSchema
      .omit({
        blockHash: true,
        blockNumber: true,
        blockTimestamp: true,
      })
      .partial()
      .required({ hash: true })
      .extend({
        blobs: z.array(
          baseBlobSchema.partial().required({ versionedHash: true })
        ),
      })
  ),
});

export type ResponseBlock = z.input<typeof responseBlockSchema>;

export function toResponseBlock(
  prismaBlock: CompletePrismaBlock
): ResponseBlock {
  const transactions = prismaBlock.transactions.map(
    ({ hash, blobs, ...prismaTx }) => ({
      hash,
      ...(isFullyDefined(prismaTx)
        ? {
            ...prismaTx,
            ...normalizePrismaTransactionFields(prismaTx),
            ...deriveTransactionFields({
              ...prismaTx,
              blobGasPrice: prismaBlock.blobGasPrice,
            }),
          }
        : {}),
      blobs: blobs.map(({ blobHash, blob }) =>
        blob
          ? normalizePrismaBlobFields({ versionedHash: blobHash, ...blob })
          : { versionedHash: blobHash }
      ),
    })
  );

  return {
    ...prismaBlock,
    transactions,
  };
}

export type BlockId = "hash" | "number" | "slot";
export type BlockIdField =
  | { type: "hash"; value: string }
  | { type: "number"; value: number }
  | { type: "slot"; value: number };

function buildBlockWhereClause(
  { type, value }: BlockIdField,
  filters: Filters
): Prisma.BlockWhereInput {
  switch (type) {
    case "hash": {
      return { hash: value };
    }
    case "number": {
      return { number: value, transactionForks: filters.blockType };
    }
    case "slot": {
      return { slot: value, transactionForks: filters.blockType };
    }
  }
}

export async function fetchBlock(
  blockId: BlockIdField,
  {
    blobStorageManager,
    prisma,
    filters,
    expands,
  }: {
    blobStorageManager: BlobStorageManager;
    prisma: BlobscanPrismaClient;
    filters: Filters;
    expands: Expands;
  }
): Promise<CompletePrismaBlock | undefined> {
  const select = createBlockSelect(expands);
  const where = buildBlockWhereClause(blockId, filters);

  const prismaBlock = (await prisma.block.findFirst({
    select,
    where,
  })) as unknown as CompletePrismaBlock | null;

  if (!prismaBlock) {
    return;
  }

  if (expands.blobData) {
    await Promise.all(
      prismaBlock.transactions
        .flatMap((tx) => tx.blobs)
        .map(async ({ blobHash, blob }) => {
          const dataStorageReferences = blob?.dataStorageReferences;
          if (dataStorageReferences) {
            const data = await retrieveBlobData(blobStorageManager, {
              dataStorageReferences,
              versionedHash: blobHash,
            });

            blob.data = data;
          }
        })
    );
  }

  return prismaBlock;
}
