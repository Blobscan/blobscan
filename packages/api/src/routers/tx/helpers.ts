import type { Prisma } from "@blobscan/db";
import { z } from "@blobscan/zod";

import type {
  ExpandedBlob,
  ExpandedBlock,
  Expands,
} from "../../middlewares/withExpands";
import type { Prettify } from "../../types";
import {
  deriveTransactionFields,
  normalizePrismaBlobFields,
  normalizePrismaTransactionFields,
} from "../../utils/transformers";
import {
  baseTransactionSchema,
  baseBlobSchema,
  baseBlockSchema,
} from "../../zod-schemas";

const transactionSelect = {
  hash: true,
  fromId: true,
  toId: true,
  blobGasUsed: true,
  blobAsCalldataGasUsed: true,
  gasPrice: true,
  maxFeePerBlobGas: true,
  blockHash: true,
  blockNumber: true,
  blockTimestamp: true,
  index: true,
  decodedFields: true,
  from: {
    select: {
      address: true,
      rollup: true,
    },
  },
} satisfies Prisma.TransactionSelect;

type PrismaTransaction = Prisma.TransactionGetPayload<{
  select: typeof transactionSelect;
}>;

export type CompletePrismaTransaction = Prettify<
  PrismaTransaction & {
    decodedFields: NonNullable<PrismaTransaction["decodedFields"]>;
    block: Prettify<
      Partial<ExpandedBlock> & {
        blobGasPrice: ExpandedBlock["blobGasPrice"];
      }
    >;
    blobs: Prettify<
      { blobHash: string } & {
        blob?: ExpandedBlob;
      }
    >[];
  }
>;
export function createTransactionSelect(expands: Expands) {
  const blockExpand = expands.block?.select;
  const blobExpand = expands.blob;

  return {
    ...transactionSelect,
    block: {
      select: {
        ...(blockExpand ?? {}),
        blobGasPrice: true,
      },
    },
    blobs: {
      select: {
        blobHash: true,
        ...(blobExpand ? { blob: blobExpand } : {}),
      },
      orderBy: {
        index: "asc",
      },
    },
  } satisfies Prisma.TransactionSelect;
}

export const responseTransactionSchema = baseTransactionSchema.extend({
  block: baseBlockSchema
    .omit({
      hash: true,
      number: true,
      timestamp: true,
    })
    .partial()
    .required({
      blobGasPrice: true,
    }),
  blobs: z.array(baseBlobSchema.partial().required({ versionedHash: true })),
});

export type ResponseTransaction = z.input<typeof responseTransactionSchema>;

export function toResponseTransaction(
  prismaTx: CompletePrismaTransaction
): ResponseTransaction {
  const { blobs: blobsOnTxs, block } = prismaTx;
  const normalizedFields = normalizePrismaTransactionFields(prismaTx);
  const derivedFields = deriveTransactionFields({
    ...prismaTx,
    blobGasPrice: block.blobGasPrice,
  });
  const blobs = blobsOnTxs.map(({ blobHash, blob }) =>
    blob
      ? normalizePrismaBlobFields({ versionedHash: blobHash, ...blob })
      : { versionedHash: blobHash }
  );

  return {
    ...prismaTx,
    ...normalizedFields,
    ...derivedFields,
    blobs,
  };
}
