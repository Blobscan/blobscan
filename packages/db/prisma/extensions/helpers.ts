import { Prisma } from "@prisma/client";

export type ZeroOpResult = { count: number }[];

export const helpersExtension = Prisma.defineExtension((prisma) =>
  prisma.$extends({
    name: "Helper Operations",
    model: {
      $allModels: {
        zero() {
          return prisma.$queryRaw<ZeroOpResult>`SELECT 0 as count`;
        },
      },
      block: {
        findLatest() {
          return prisma.block.findFirst({
            where: {
              transactionForks: {
                none: {},
              },
            },
            orderBy: { number: "desc" },
          });
        },
      },
    },
  })
);
