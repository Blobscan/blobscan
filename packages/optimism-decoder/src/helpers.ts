import axios from "axios";
import fs from "fs";
import path from "path";

import type { OptimismDecodedData } from "./decoder";

/**
 * Read the binary file and split it into chunks of the specified size.
 * @param buffer - The binary data from the file.
 * @param chunkSize - The size of each chunk.
 * @returns An array of chunks.
 */
export function chunks(buffer: Uint8Array, chunkSize: number): Uint8Array[] {
  const result = new Array(Math.ceil(buffer.length / chunkSize));
  for (let i = 0; i < buffer.length; i += chunkSize) {
    result[i / chunkSize] = buffer.slice(i, i + chunkSize);
  }
  return result;
}

/**
 * Convert the byte array to a number.
 * @param bytes - The array of bytes to convert.
 * @returns The number representation of the bytes.
 */
export function bytesToNumber(bytes: Uint8Array): number {
  return bytes.reduce(
    (acc, byte, index) => acc + (byte << (8 * (bytes.length - index - 1))),
    0
  );
}

type FetchResult = {
  versionedHashes: string[];
  blockNumber: number;
};

/**
 * Function to fetch data from Transactions API endpoint and extract versionedHash values.
 * @param transactionId - The transaction id.
 * @returns An array of versionedHash values.
 */
export async function fetchAndExtractVersionedHashes(
  transactionId: string
): Promise<FetchResult> {
  const endpoint = `https://api.blobscan.com/transactions/${transactionId}`;
  try {
    const response = await axios.get(endpoint);
    const data = response.data;

    const blockNumber: number = data.blockNumber;

    // Extract the versionedHash values from the blobs array
    const versionedHashes: string[] = data.blobs.map(
      (blob: { versionedHash: string }) => blob.versionedHash
    );

    console.log(`[${data.blockTimestamp}] Block: ${blockNumber}`);
    return {
      versionedHashes,
      blockNumber,
    };
  } catch (error) {
    console.error("Error fetching or processing data:", error);
    return { versionedHashes: [], blockNumber: 0 };
  }
}

/**
 * Function to download data from multiple endpoints based on the versionedHash list and save it to disk.
 * @param hashes - The list of versionedHash values.
 */
export async function downloadAndSaveFiles(
  blockNumber: number,
  transactionId: string,
  hashes: string[]
): Promise<string> {
  const targetFolder = path.join(__dirname, "blobs", "tx", transactionId);
  if (!fs.existsSync(targetFolder)) {
    fs.mkdirSync(targetFolder, { recursive: true });
    console.log(`Directory created: ${targetFolder}`);
  }

  const filePaths: string[] = [];

  let index = 0;
  for (const hash of hashes) {
    const url = `https://api.blobscan.com/blobs/${hash}`;
    try {
      const response = await axios.get(url);
      const blobHexData = response.data.data;
      const buffer = Buffer.from(blobHexData.slice(2), "hex"); // Remove the "0x" prefix before converting

      const fileName = `${index}.bin`;
      const filePath = path.join(targetFolder, fileName);

      fs.writeFileSync(filePath, buffer);
      filePaths.push(filePath);
      index++;
    } catch (error) {
      console.error(`Failed to download data for ${hash}:`, error);
    }
  }
  const filename = combineFiles(blockNumber, transactionId, filePaths);
  return filename;
}

/**
 * Function to combine multiple binary files into a single file.
 * @param filePaths - Array of file paths to combine.
 */
function combineFiles(
  blockNumber: number,
  transactionId: string,
  filePaths: string[]
): string {
  const outputDir = path.join(__dirname, "blobs", "tx", transactionId);
  const outputFilePath = path.join(outputDir, `final_${blockNumber}.bin`);

  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
    console.log(`Directory created: ${outputDir}`);
  }

  const fileDescriptor = fs.openSync(outputFilePath, "w");

  try {
    for (const filePath of filePaths) {
      const fileBuffer = fs.readFileSync(filePath);
      fs.writeSync(fileDescriptor, fileBuffer);
    }
  } catch (error) {
    const error_ = error as Error;

    console.error("Error combining files:", error_.message);
  } finally {
    fs.closeSync(fileDescriptor);
  }

  return outputFilePath;
}

/**
 * Function to save OptimismDecodedData to a JSON file (decoded.json).
 * @param data - The OptimismDecodedData object to save.
 * @param transactionId - The transaction id to include in the filename.
 */
export function saveOptimismDecodedData(
  data: OptimismDecodedData,
  transactionId: string
): void {
  const outputDir = path.join(__dirname, "blobs", "tx", transactionId);
  const outputFilePath = path.join(outputDir, "decoded.json");

  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
    console.log(`Directory created: ${outputDir}`);
  }

  const jsonData = JSON.stringify(data, null, 2);

  fs.writeFileSync(outputFilePath, jsonData, "utf8");
  console.log(`OptimismDecodedData saved to ${outputFilePath}`);
}
