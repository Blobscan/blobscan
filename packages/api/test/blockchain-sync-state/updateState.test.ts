import { TRPCError } from "@trpc/server";
import { beforeAll, beforeEach, describe, expect, it } from "vitest";

import type { BlockchainSyncState } from "@blobscan/db";
import { testValidError } from "@blobscan/test";

import type { TRPCContext } from "../../src";
import { createTestContext, unauthorizedRPCCallTest } from "../helpers";
import { createBlockchainSyncStateCaller } from "./caller";
import type { BlockchainSyncStateCaller } from "./caller";

describe("updateState", () => {
  let nonAuthorizedCaller: BlockchainSyncStateCaller;
  let authorizedCaller: BlockchainSyncStateCaller;
  let authorizedContext: TRPCContext;

  beforeAll(async () => {
    const ctx = await createTestContext();

    authorizedContext = await createTestContext({
      apiClient: "indexer",
    });

    nonAuthorizedCaller = createBlockchainSyncStateCaller(ctx);
    authorizedCaller = createBlockchainSyncStateCaller(authorizedContext);
  });

  describe("when authorized", () => {
    describe("when updating blockchain sync state", () => {
      let prevBlockchainSyncState: BlockchainSyncState;

      beforeEach(async () => {
        const state =
          await authorizedContext.prisma.blockchainSyncState.findFirst();

        if (!state) {
          throw new Error("Blockchain sync state should exist");
        }

        prevBlockchainSyncState = state;
      });

      it("should update the last finalized block correctly", async () => {
        const newLastFinalizedBlock = 2000;

        await authorizedCaller.updateState({
          lastFinalizedBlock: newLastFinalizedBlock,
        });

        const afterBlockchainSyncState =
          await authorizedContext.prisma.blockchainSyncState.findFirst();

        expect(afterBlockchainSyncState).toMatchObject({
          ...prevBlockchainSyncState,
          lastFinalizedBlock: newLastFinalizedBlock,
        });
      });

      it("should update the last lower synced slot correctly", async () => {
        const newLastLowerSyncedSlot = 2000;

        await authorizedCaller.updateState({
          lastLowerSyncedSlot: newLastLowerSyncedSlot,
        });

        const afterBlockchainSyncState =
          await authorizedContext.prisma.blockchainSyncState.findFirst();

        expect(afterBlockchainSyncState).toMatchObject({
          ...prevBlockchainSyncState,
          lastLowerSyncedSlot: newLastLowerSyncedSlot,
        });
      });

      it("should update the last upper synced slot correctly", async () => {
        const newLastUpperSyncedSlot = 2000;

        await authorizedCaller.updateState({
          lastUpperSyncedSlot: newLastUpperSyncedSlot,
        });

        const afterBlockchainSyncState =
          await authorizedContext.prisma.blockchainSyncState.findFirst();

        expect(afterBlockchainSyncState).toMatchObject({
          ...prevBlockchainSyncState,
          lastUpperSyncedSlot: newLastUpperSyncedSlot,
        });
      });

      it("should update the last upper synced block root correctly", async () => {
        const newLastUpperSyncedBlockRoot = "0x".padEnd(66, "1");

        await authorizedCaller.updateState({
          lastUpperSyncedBlockRoot: newLastUpperSyncedBlockRoot,
        });

        const afterBlockchainSyncState =
          await authorizedContext.prisma.blockchainSyncState.findFirst();

        expect(afterBlockchainSyncState).toMatchObject({
          ...prevBlockchainSyncState,
          lastUpperSyncedBlockRoot: newLastUpperSyncedBlockRoot,
        });
      });

      it("should update the last upper synced block slot correctly", async () => {
        const newLastUpperSyncedBlockSlot = 2000;

        await authorizedCaller.updateState({
          lastUpperSyncedBlockSlot: newLastUpperSyncedBlockSlot,
        });

        const afterBlockchainSyncState =
          await authorizedContext.prisma.blockchainSyncState.findFirst();

        expect(afterBlockchainSyncState).toMatchObject({
          ...prevBlockchainSyncState,
          lastUpperSyncedBlockSlot: newLastUpperSyncedBlockSlot,
        });
      });

      it("should update all fields correctly", async () => {
        const newBlockchainSyncState = {
          lastFinalizedBlock: 2001,
          lastLowerSyncedSlot: 30,
          lastUpperSyncedSlot: 560,
          lastUpperSyncedBlockRoot: "0x".padEnd(66, "1"),
          lastUpperSyncedBlockSlot: 2000,
        };

        await authorizedCaller.updateState(newBlockchainSyncState);

        const afterBlockchainSyncState =
          await await authorizedContext.prisma.blockchainSyncState.findFirst();

        expect(afterBlockchainSyncState).toMatchObject(newBlockchainSyncState);
      });
    });

    testValidError(
      "should fail when trying to update last finalized slot to an invalid slot",
      async () => {
        await authorizedCaller.updateState({
          lastFinalizedBlock: -1,
        });
      },
      TRPCError,
      {
        checkCause: true,
      }
    );

    testValidError(
      "should fail when trying to update last lower synced slot to an invalid slot",
      async () => {
        await authorizedCaller.updateState({
          lastLowerSyncedSlot: -1,
        });
      },
      TRPCError,
      {
        checkCause: true,
      }
    );

    testValidError(
      "should fail when trying to update last upper synced slot to an invalid slot",
      async () => {
        await authorizedCaller.updateState({
          lastUpperSyncedSlot: -1,
        });
      },
      TRPCError,
      {
        checkCause: true,
      }
    );

    testValidError(
      "should fail when trying to update last upper synced block root to an invalid hash",
      async () => {
        await authorizedCaller.updateState({
          lastUpperSyncedBlockRoot: "invalid hash",
        });
      },
      TRPCError,
      {
        checkCause: true,
      }
    );

    testValidError(
      "should fail when trying to update last upper synced block slot to an invalid slot",
      async () => {
        await authorizedCaller.updateState({
          lastUpperSyncedBlockSlot: -1,
        });
      },
      TRPCError,
      {
        checkCause: true,
      }
    );

    it("should fail when trying to update last lower synced slot to a value greater than last upper synced slot", async () => {
      const newState = {
        lastLowerSyncedSlot: 1001,
        lastUpperSyncedSlot: 1000,
      };

      await expect(
        authorizedCaller.updateState(newState)
      ).rejects.toThrowErrorMatchingInlineSnapshot(
        '"Last lower synced slot must be less than or equal to last upper synced slot"'
      );
    });
  });

  unauthorizedRPCCallTest(() =>
    nonAuthorizedCaller.updateState({
      lastLowerSyncedSlot: 1001,
    })
  );
});
