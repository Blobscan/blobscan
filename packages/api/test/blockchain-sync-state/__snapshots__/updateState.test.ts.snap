// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`updateState > when authorized > should fail when trying to update last finalized slot to an invalid slot 1`] = `
"[
  {
    \\"code\\": \\"too_small\\",
    \\"minimum\\": 0,
    \\"type\\": \\"number\\",
    \\"inclusive\\": true,
    \\"exact\\": false,
    \\"message\\": \\"Number must be greater than or equal to 0\\",
    \\"path\\": [
      \\"lastFinalizedBlock\\"
    ]
  }
]"
`;

exports[`updateState > when authorized > should fail when trying to update last finalized slot to an invalid slot 2`] = `
[ZodError: [
  {
    "code": "too_small",
    "minimum": 0,
    "type": "number",
    "inclusive": true,
    "exact": false,
    "message": "Number must be greater than or equal to 0",
    "path": [
      "lastFinalizedBlock"
    ]
  }
]]
`;

exports[`updateState > when authorized > should fail when trying to update last lower synced slot to an invalid slot 1`] = `
"[
  {
    \\"code\\": \\"too_small\\",
    \\"minimum\\": 0,
    \\"type\\": \\"number\\",
    \\"inclusive\\": true,
    \\"exact\\": false,
    \\"message\\": \\"Number must be greater than or equal to 0\\",
    \\"path\\": [
      \\"lastLowerSyncedSlot\\"
    ]
  }
]"
`;

exports[`updateState > when authorized > should fail when trying to update last lower synced slot to an invalid slot 2`] = `
[ZodError: [
  {
    "code": "too_small",
    "minimum": 0,
    "type": "number",
    "inclusive": true,
    "exact": false,
    "message": "Number must be greater than or equal to 0",
    "path": [
      "lastLowerSyncedSlot"
    ]
  }
]]
`;

exports[`updateState > when authorized > should fail when trying to update last upper synced block root to an invalid hash 1`] = `
"[
  {
    \\"validation\\": \\"regex\\",
    \\"code\\": \\"invalid_string\\",
    \\"message\\": \\"Invalid hexadecimal string\\",
    \\"path\\": [
      \\"lastUpperSyncedBlockRoot\\"
    ]
  },
  {
    \\"code\\": \\"too_small\\",
    \\"minimum\\": 66,
    \\"type\\": \\"string\\",
    \\"inclusive\\": true,
    \\"exact\\": true,
    \\"message\\": \\"Invalid block hash length\\",
    \\"path\\": [
      \\"lastUpperSyncedBlockRoot\\"
    ]
  }
]"
`;

exports[`updateState > when authorized > should fail when trying to update last upper synced block root to an invalid hash 2`] = `
[ZodError: [
  {
    "validation": "regex",
    "code": "invalid_string",
    "message": "Invalid hexadecimal string",
    "path": [
      "lastUpperSyncedBlockRoot"
    ]
  },
  {
    "code": "too_small",
    "minimum": 66,
    "type": "string",
    "inclusive": true,
    "exact": true,
    "message": "Invalid block hash length",
    "path": [
      "lastUpperSyncedBlockRoot"
    ]
  }
]]
`;

exports[`updateState > when authorized > should fail when trying to update last upper synced block slot to an invalid slot 1`] = `
"[
  {
    \\"code\\": \\"too_small\\",
    \\"minimum\\": 0,
    \\"type\\": \\"number\\",
    \\"inclusive\\": true,
    \\"exact\\": false,
    \\"message\\": \\"Number must be greater than or equal to 0\\",
    \\"path\\": [
      \\"lastUpperSyncedBlockSlot\\"
    ]
  }
]"
`;

exports[`updateState > when authorized > should fail when trying to update last upper synced block slot to an invalid slot 2`] = `
[ZodError: [
  {
    "code": "too_small",
    "minimum": 0,
    "type": "number",
    "inclusive": true,
    "exact": false,
    "message": "Number must be greater than or equal to 0",
    "path": [
      "lastUpperSyncedBlockSlot"
    ]
  }
]]
`;

exports[`updateState > when authorized > should fail when trying to update last upper synced slot to an invalid slot 1`] = `
"[
  {
    \\"code\\": \\"too_small\\",
    \\"minimum\\": 0,
    \\"type\\": \\"number\\",
    \\"inclusive\\": true,
    \\"exact\\": false,
    \\"message\\": \\"Number must be greater than or equal to 0\\",
    \\"path\\": [
      \\"lastUpperSyncedSlot\\"
    ]
  }
]"
`;

exports[`updateState > when authorized > should fail when trying to update last upper synced slot to an invalid slot 2`] = `
[ZodError: [
  {
    "code": "too_small",
    "minimum": 0,
    "type": "number",
    "inclusive": true,
    "exact": false,
    "message": "Number must be greater than or equal to 0",
    "path": [
      "lastUpperSyncedSlot"
    ]
  }
]]
`;
